gear-contract-libraries/multitoken/mtk_core.rs

```diff
@@ -0,0 +1,326 @@
+        self.set_balance(account, id, prev_balance.saturating_add(amount));
+    }
+
+    fn mint(&mut self, account: &ActorId, id: &TokenId, amount: u128, meta: Option<TokenMetadata>) {
+        self._mint(account, id, amount, meta);
+        msg::reply(
+            MTKEvent::TransferSingle(TransferSingleReply {
+                operator: msg::source(),
+                from: ZERO_ID,
+                to: *account,
+                id: *id,
+                amount,
+            }),
+            0,
+        )
+        .unwrap();
+    }
+
+    fn mint_batch(
+        &mut self,
+        account: &ActorId,
+        ids: &[TokenId],
+        amounts: &[u128],
+        meta: Vec<Option<TokenMetadata>>,
+    ) {
+        if account == &ZERO_ID {
+            panic!("MTK: Mint to zero address")
+        }
+
+        if ids.len() != amounts.len() {
+            panic!("MTK: ids and amounts length mismatch")
+        }
+
+        meta.into_iter()
+            .enumerate()
+            .for_each(|(i, meta)| self._mint(account, &ids[i], amounts[i], meta));
+
+        // TODO: rewrites
+        msg::reply(
+            MTKEvent::TransferBatch {
+                operator: msg::source(),
+                from: ZERO_ID,
+                to: *account,
+                ids: ids.to_vec(),
+                values: amounts.to_vec(),
+            },
+            0,
+        )
+        .unwrap();
+    }
+
+    fn _burn(&mut self, id: &TokenId, amount: u128) {
+        let owner = &msg::source();
+        self.assert_can_burn(owner, id, amount);
+        self.set_balance(
+            &msg::source(),
+            id,
+            self.get_balance(owner, id).saturating_sub(amount),
+        );
+    }
+
+    fn burn(&mut self, id: &TokenId, amount: u128) {
+        self._burn(id, amount);
+        msg::reply(
+            MTKEvent::TransferSingle(TransferSingleReply {
+                operator: msg::source(),
+                from: msg::source(),
+                to: ZERO_ID,
+                id: *id,
+                amount,
+            }),
+            0,
+        )
+        .unwrap();
+    }
+
+    fn burn_batch(&mut self, ids: &[TokenId], amounts: &[u128]) {
+        if ids.len() != amounts.len() {
+            panic!("MTK: ids and amounts length mismatch")
+        }
+
+        for (id, amount) in ids.iter().zip(amounts) {
+            self.assert_can_burn(&msg::source(), id, *amount);
+        }
+
+        ids.iter()
+            .enumerate()
+            .for_each(|(i, id)| self._burn(id, amounts[i]));
+
+        msg::reply(
+            MTKEvent::TransferBatch {
+                operator: msg::source(),
+                from: msg::source(),
+                to: ZERO_ID,
+                ids: ids.to_vec(),
+                values: amounts.to_vec(),
+            },
+            0,
+        )
+        .unwrap();
+    }
+    fn _transfer_from(&mut self, from: &ActorId, to: &ActorId, id: &TokenId, amount: u128) {
```

Comment:
Not too good a name in Rust.
```suggestion
    fn transfer_from_impl(&mut self, from: &ActorId, to: &ActorId, id: &TokenId, amount: u128) {
```